Add more tests

baru64 opened the pull request.

baru64 committed:
Add UserViewsTest

baru64 committed:
Add WebpageViewsTest, TemplateViewsTest, CommentViewsTest

baru64 committed:
Fix CommentCreateView redirect

baru64 committed:
Fix test_comment_delete

A5rocks commented on jolly-jellyfish/src/page_maker/fixtures/templates.yaml:
```diff
@@ -0,0 +1,14 @@
+    name: doghot
+    date_created: 2020-08-02 19:51:03.603472+00:00
+    author: 1
+    style_sheet: styles/test2.css
```
Should we maybe make a separate directory for test files?

baru64 replied:
We don't need real files, at least for now and I don't see any need for that, if the url for image/file is present it's enough.

A5rocks commented on jolly-jellyfish/src/page_maker/fixtures/webpages.yaml:
```diff
@@ -0,0 +1,17 @@
+    user_text_1: asdafgssag
+    user_text_2: sdgagdasd
+    user_text_3: xxcvxcvbxcvbxcvbxcvbbcvbxc
+    user_image_1: images/admin/img1.jpg
+    user_image_2: images/admin/img2.jpg
+    user_image_3: images/admin/img3.jpg
+    user_image_4: images/admin/img4.jpg
```
Are these default images that come with a normal Django install? (Just wondering... Since otherwise, let's move them under a separate directory for test files :-)

baru64 replied:
Just a placeholder since we don't need real files.

A5rocks commented on jolly-jellyfish/src/page_maker/tests.py:
```diff
@@ -20,6 +27,11 @@ def setUp(self):
+    def test_user_exists(self):
+        self.assertEqual(User.objects.count(), 1)
+        user = User.objects.get(pk=1)
+        self.assertEqual(user.username, USERNAME)
```
Should we also check the email?

A5rocks commented on jolly-jellyfish/src/page_maker/tests.py:
```diff
@@ -38,13 +50,115 @@ def test_register_and_home(self):
+
+    def test_user_detail(self):
+        user = User.objects.get(pk=1)
+        res = self.client.get('/users/1/')
```
Is this trailing slash required? If so, then we might want to fix our URLCONF.

A5rocks commented on jolly-jellyfish/src/page_maker/tests.py:
```diff
@@ -38,13 +50,115 @@ def test_register_and_home(self):
+        self.assertEqual(user.last_name, 'lemon')
+
+    def test_user_delete(self):
+        self.client.login(username=USERNAME, password=PASSWORD)
```
We could probably move this line to the `setUp`, given that `test_user_detail` shouldn't be dependent on a user being signed in.

A5rocks commented on jolly-jellyfish/src/page_maker/tests.py:
```diff
@@ -38,13 +50,115 @@ def test_register_and_home(self):
+        self.assertEqual(res.context['viewed_user'], user)
+        self.assertIsInstance(res.context['user_pages'][0], Webpage)
+
+    def test_user_update(self):
```
Should we test that a non-signed-in user cannot update a user? That sounds like some logic we want to make sure works.

A5rocks commented on jolly-jellyfish/src/page_maker/tests.py:
```diff
@@ -38,13 +50,115 @@ def test_register_and_home(self):
+        self.client.login(username=USERNAME, password=PASSWORD)
+        res = self.client.get('/pages/new')
+        self.assertEqual(res.status_code, 200)
+        self.assertTemplateUsed(res, template_name='page_maker/webpage_create.html')
```
Are we sure we want to hardcode the template used? This might make for a flakey test (as in: a test we will need to change eventually if we rename templates or something).

Though given our dependence on certain items being in the `res.context` this might not be so important...

A5rocks reviewed (commented):
Looks good! I have a couple questions, but I didn't see anything bad.

baru64 committed:
Improve tests

Changes:
  - remove trailing slash in urls
  - swap hardcoded paths with reverse()
  - remove template tests

A5rocks commented on jolly-jellyfish/src/page_maker/tests.py:
```diff
@@ -55,14 +57,18 @@ class UserViewsTest(TestCase):
+        res = self.client.post(reverse('user-update', kwargs={'pk': 1}), {
+            'first_name': 'john', 'last_name': 'lemon', 'email': 'jl@sth.com'
+        })
+        self.assertEqual(res.status_code, 403)
```
I think you messed up the ordering?

A5rocks replied:
NVM I missed the `{'pk': 1}`.

A5rocks reviewed (commented):
There's just this one thing I'm not sure of...

Nevermind, it was just me missing part of the code...
